src/Monitor/Tracing/Zipkin.hs

```diff
@@ -348,24 +348,28 @@ clientSpanWith f = outgoingSpan "CLIENT" (Endo f)
 producerSpanWith :: MonadTrace m => (Builder -> Builder) -> Name -> (Maybe B3 -> m a) -> m a
 producerSpanWith f = outgoingSpan "PRODUCER" (Endo f)
 
-incomingSpan :: MonadTrace m => Text -> Endo Builder -> B3 -> m a -> m a
-incomingSpan kind endo b3 actn =
-  let bldr = appEndo (importB3 b3 <> insertTag kindKey kind <> endo) $ builder ""
+incomingSpan :: MonadTrace m => Text -> Endo Builder -> m a -> m a
+incomingSpan kind endo actn =
+  let bldr = appEndo (insertTag kindKey kind <> endo) $ builder ""
   in trace bldr actn
 
 -- | Generates a child span with @SERVER@ kind. The client's 'B3' should be provided as input,
 -- for example parsed using 'b3FromHeaders'.
 serverSpan :: MonadTrace m => B3 -> m a -> m a
 serverSpan = serverSpanWith id
 
--- | Generates a server span, optionally modifying the span's builder. This can be useful in
+-- | Generates a child span with @SERVER@ kind, optionally modifying the span's builder. This can be useful in
 -- combination with 'addEndpoint' if the remote client does not have tracing enabled.
+-- The clients's 'B3' should be provided as input.
+-- Client and server annotations go on the same span - it means that they share their span ID.
 serverSpanWith :: MonadTrace m => (Builder -> Builder) -> B3 -> m a -> m a
-serverSpanWith f = incomingSpan "SERVER" (Endo f)
+serverSpanWith f b3 = incomingSpan "SERVER" (importB3 b3 <> Endo (\bldr -> f $ bldr { builderSpanID = Just (b3SpanID b3) }))
```

Comment:
Call `importB3` inside `incomingSpan` to avoid duplication?

Reply:
Hope it is o.k. that I've inserted the B3 parameter behind kind - my thought was that the kind is more important and that it leads to more natural sequence. E.g. incomingSpan (of kind) "SERVER" (with span) b3 [rest].